pinot-core/src/main/java/org/apache/pinot/server/realtime/ServerSegmentCompletionProtocolHandler.java

```diff
@@ -131,7 +140,15 @@ public ServerSegmentCompletionProtocolHandler(ServerMetrics serverMetrics, Strin
       return SegmentCompletionProtocol.RESP_NOT_SENT;
     }
 
-    return uploadSegment(url, params.getSegmentName(), segmentTarFile);
+    Server2ControllerSegmentUploader segmentUploader= null;
+    try {
+      segmentUploader = new Server2ControllerSegmentUploader(LOGGER,
```

Comment:
I looked up the code, and it seems that we sanitize the url returned by the controller, but then we pass strings around. We need to clean this up, but it can be in a later PR.